File: src/haskell/sp/Signal/Padding.hs
{-# LANGUAGE  ConstraintKinds, RankNTypes #-}

module Signal.Padding
(Padding(..)
,Extensions(..)
,PaddingMethod(ZERO,PERIODIC,SYMETRIC)
) where

import Signal.Types

data Extensions a = Extensions{left::[a]
                              ,right::[a]}
instance (Show a)=>Show (Extensions a) where
    show Extensions{left=l,right=r} = "["++ show l ++ "," ++ show r ++ "]" ++ "\n"

class Padding f where
    pad::SigNum a=>f->(Int->[a]->Extensions a)

data PaddingMethod=ZERO|PERIODIC|SYMETRIC deriving(Eq)

type GetExtensions a = SigNum a=>Int->[a]->Extensions a

zero::GetExtensions a
zero n a = Extensions{left=replicate n 0, right=replicate n 0}

symetric::GetExtensions a
symetric n a = Extensions{left=lr,right=rr} where
    lr = reverse $ take n a
    rr = take n $ reverse a

periodic::GetExtensions a
periodic n a = Extensions{left=lr,right=rr} where
    lr = reverse $ take n  $ reverse a
    rr = take n a

instance Padding PaddingMethod where
    pad ZERO = zero
    pad PERIODIC = periodic
    pad SYMETRIC = symetric

File: src/python/Signal/Padding.py
from typing import NewType
from enum import Enum
from Signal.Types import sig_0

Extensions = NewType('Extensions',tuple[sig_0,sig_0])
PaddingMethodNames=Enum('PaddingMethod',['ZERO','PERIODIC','SYMETRIC'])

def zero(n:int,v:sig_0)->Extensions:
    if n<0 or n>len(v): raise ValueError("Illegal Input")
    else: return Extensions(([0.0]*n,[0.0]*n))

def periodic(n:int,v:sig_0)->Extensions:
    if n<0 or n>len(v): raise ValueError("Illegal Input")
    else: return Extensions((v[-n:],v[:n]))

def symetric(n:int,v:sig_0)->Extensions:
    if n<0 or n>len(v): raise ValueError("Illegal Input")
    else: return Extensions((v[:n][::-1],v[-n:][::-1]))

padding={name:method for (name,method) in zip(PaddingMethodNames,(zero,periodic,symetric))}
